### ISSUE TITLE
cli: `cockroach debug range-data` crashes

### ISSUE THREAD

**Issue by smcvey** [cockroachdb/cockroach#69769]
I ran `cockroach debug range-data cockroach-data 37` which caused a panic and produced a crash output.

I created a new table and inserted some data:

```
create database shaun;
use shaun;

create table a (
        a int primary key default unique_rowid(),
        b varchar,
        c uuid default gen_random_uuid()
);

insert into a select a from generate_series(1, 10000) g(a);
```

Then I plucked out the range id:

```
root@:26257/shaun> show ranges from database shaun;
  table_name | start_key | end_key | range_id | range_size_mb | lease_holder | lease_holder_locality | replicas | replica_localities
-------------+-----------+---------+----------+---------------+--------------+-----------------------+----------+---------------------
  a          | NULL      | NULL    |       37 |      0.409636 |            1 |                       | {1,2,3}  | {,,}
(1 row)
```

Then I stopped the server and then ran `cockroach debug range-data cockroach-data 37`.

This produced this output:

```
$ cockroach debug range-data cockroach-data/ 37
0,0 /Local/RangeID/37/r/RangeLastGC (0x0169ad726c67632d00): %!s(PANIC=String method: field raftpb.Entry.Term has invalid type: got uint64, want pointer): EMPTY

0,0 /Local/RangeID/37/r/RangeAppliedState (0x0169ad727261736b00): raft_applied_index:112 lease_applied_index:104 range_stats:<last_update_nanos:1630592726775912378 live_bytes:409636 live_count:10000 key_bytes:189636 key_count:10000 val_bytes:220000 val_count:10000 sys_bytes:183 sys_count:4 > raft_closed_timestamp:<wall_time:1630592723776049298 > 
0,0 /Local/RangeID/37/r/RangeLease (0x0169ad72726c6c2d00): repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1630592491.754051460,0
0,0 /Local/RangeID/37/r/RangeVersion (0x0169ad727276657200): 21.1-124
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1  a panic has occurred!
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +panic: field raftpb.HardState.term has invalid type: got uint64, want pointer
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +(1) attached stack trace
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  -- stack trace:
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | runtime.gopanic
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/usr/local/go/src/runtime/panic.go:969
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | google.golang.org/protobuf/internal/impl.fieldInfoForScalar
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/protobuf/internal/impl/message_reflect_field.go:228
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | google.golang.org/protobuf/internal/impl.(*MessageInfo).makeKnownFieldsFunc
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/protobuf/internal/impl/message_reflect.go:67
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | google.golang.org/protobuf/internal/impl.(*MessageInfo).makeReflectFuncs
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/protobuf/internal/impl/message_reflect.go:36
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | google.golang.org/protobuf/internal/impl.(*MessageInfo).initOnce
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/protobuf/internal/impl/message.go:90
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | google.golang.org/protobuf/internal/impl.(*MessageInfo).init
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/protobuf/internal/impl/message.go:72
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | google.golang.org/protobuf/internal/impl.(*messageReflectWrapper).Has
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/protobuf/internal/impl/message_reflect_gen.go:185
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/golang/protobuf/proto.(*textWriter).writeMessage
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/golang/protobuf/proto/text_encode.go:278
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/golang/protobuf/proto.(*TextMarshaler).marshal
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/golang/protobuf/proto/text_encode.go:86
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/golang/protobuf/proto.(*TextMarshaler).Text
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/golang/protobuf/proto/text_encode.go:44
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/golang/protobuf/proto.CompactTextString
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/golang/protobuf/proto/text_encode.go:106
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | go.etcd.io/etcd/raft/v3/raftpb.(*HardState).String
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/vendor/go.etcd.io/etcd/raft/v3/raftpb/raft.pb.go:443
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.tryRangeIDKey
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/debug_print.go:311
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.SprintKeyValue
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/debug_print.go:68
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.PrintKeyValue
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/debug_print.go:33
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.PrintEngineKeyValue
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/debug_print.go:394
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/cli.runDebugRangeData
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/cli/debug.go:369
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/cli.MaybeDecorateGRPCError.func1
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/cli/error.go:194
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/spf13/cobra.(*Command).execute
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:826
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/spf13/cobra.(*Command).ExecuteC
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:914
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/spf13/cobra.(*Command).Execute
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:864
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/cli.Run
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:322
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/cli.doMain
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:121
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | github.com/cockroachdb/cockroach/pkg/cli.Main
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:59
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | main.main
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/cmd/cockroach/main.go:26
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | runtime.main
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/usr/local/go/src/runtime/proc.go:204
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | runtime.goexit
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +  | 	/usr/local/go/src/runtime/asm_amd64.s:1374
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +Wraps: (2) panic: field raftpb.HardState.term has invalid type: got uint64, want pointer
E210902 14:25:43.768913 1 1@util/log/logcrash/crash_reporting.go:140  [-] 1 +Error types: (1) *withstack.withStack (2) *errutil.leafError
panic: field raftpb.HardState.term has invalid type: got uint64, want pointer [recovered]
	panic: field raftpb.HardState.term has invalid type: got uint64, want pointer [recovered]
	panic: field raftpb.HardState.term has invalid type: got uint64, want pointer

goroutine 1 [running]:
github.com/cockroachdb/cockroach/pkg/util/log/logcrash.RecoverAndReportPanic(0x5aab0e0, 0xc00007a0d8, 0xc000780000)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:90 +0x9f
panic(0x441d9a0, 0xc000539aa0)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc001871500, 0x5aab0e0, 0xc00007a0d8)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:506 +0x2d3
panic(0x441d9a0, 0xc000539aa0)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9
google.golang.org/protobuf/internal/impl.fieldInfoForScalar(0x5c02500, 0xc00189eb00, 0x4345e65, 0x4, 0x0, 0x0, 0x5c02620, 0x441dba0, 0x4345e6b, 0x2d, ...)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/protobuf/internal/impl/message_reflect_field.go:228 +0x7d7
google.golang.org/protobuf/internal/impl.(*MessageInfo).makeKnownFieldsFunc(0xc0000f0dc0, 0xffffffffffffffff, 0xffffffffffffffff, 0xffffffffffffffff, 0xffffffffffffffff, 0xc00074d7d0, 0xc00074da70, 0xc00074db00, 0xc00074dbf0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/protobuf/internal/impl/message_reflect.go:67 +0x97d
google.golang.org/protobuf/internal/impl.(*MessageInfo).makeReflectFuncs(0xc0000f0dc0, 0x5c02620, 0x480d0c0, 0xffffffffffffffff, 0xffffffffffffffff, 0xffffffffffffffff, 0xffffffffffffffff, 0xc00074d7d0, 0xc00074da70, 0xc00074db00, ...)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/protobuf/internal/impl/message_reflect.go:36 +0x65
google.golang.org/protobuf/internal/impl.(*MessageInfo).initOnce(0xc0000f0dc0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/protobuf/internal/impl/message.go:90 +0x192
google.golang.org/protobuf/internal/impl.(*MessageInfo).init(...)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/protobuf/internal/impl/message.go:72
google.golang.org/protobuf/internal/impl.(*messageReflectWrapper).Has(0xc000539a20, 0x5c02500, 0xc00189eb00, 0xc00189eb00)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/protobuf/internal/impl/message_reflect_gen.go:185 +0xf9
github.com/golang/protobuf/proto.(*textWriter).writeMessage(0xc00074d410, 0x5bf4c80, 0xc000539a20, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/golang/protobuf/proto/text_encode.go:278 +0x935
github.com/golang/protobuf/proto.(*TextMarshaler).marshal(0x7d76f9c, 0x5a815a0, 0xc001943b40, 0xc000a2f0d0, 0xb, 0x10, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/golang/protobuf/proto/text_encode.go:86 +0x190
github.com/golang/protobuf/proto.(*TextMarshaler).Text(...)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/golang/protobuf/proto/text_encode.go:44
github.com/golang/protobuf/proto.CompactTextString(...)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/golang/protobuf/proto/text_encode.go:106
go.etcd.io/etcd/raft/v3/raftpb.(*HardState).String(0xc001943b40, 0xb, 0x10)
	/go/src/github.com/cockroachdb/cockroach/vendor/go.etcd.io/etcd/raft/v3/raftpb/raft.pb.go:443 +0x45
github.com/cockroachdb/cockroach/pkg/kv/kvserver.tryRangeIDKey(0xc0008795c0, 0x8, 0x40, 0x0, 0x0, 0x7f52b01a48c5, 0x19, 0x19, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/debug_print.go:311 +0x507
github.com/cockroachdb/cockroach/pkg/kv/kvserver.SprintKeyValue(0xc0008795c0, 0x8, 0x40, 0x0, 0x0, 0x7f52b01a48c5, 0x19, 0x19, 0xc000879501, 0x8, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/debug_print.go:68 +0x194
github.com/cockroachdb/cockroach/pkg/kv/kvserver.PrintKeyValue(0xc0008795c0, 0x8, 0x40, 0x0, 0x0, 0x7f52b01a48c5, 0x19, 0x19)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/debug_print.go:33 +0x5b
github.com/cockroachdb/cockroach/pkg/kv/kvserver.PrintEngineKeyValue(0xc0008795c0, 0x8, 0x40, 0x0, 0x0, 0x0, 0x7f52b01a48c5, 0x19, 0x19)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/debug_print.go:394 +0x645
github.com/cockroachdb/cockroach/pkg/cli.runDebugRangeData(0x7d4d7a0, 0xc001912f60, 0x2, 0x2, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/debug.go:369 +0x499
github.com/cockroachdb/cockroach/pkg/cli.MaybeDecorateGRPCError.func1(0x7d4d7a0, 0xc001912f60, 0x2, 0x2, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/error.go:194 +0x79
github.com/spf13/cobra.(*Command).execute(0x7d4d7a0, 0xc001912f20, 0x2, 0x2, 0x7d4d7a0, 0xc001912f20)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:826 +0x47c
github.com/spf13/cobra.(*Command).ExecuteC(0x7d4a320, 0xc00007a0d8, 0xc0017a150a, 0x10)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:914 +0x30b
github.com/spf13/cobra.(*Command).Execute(...)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:864
github.com/cockroachdb/cockroach/pkg/cli.Run(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:322
github.com/cockroachdb/cockroach/pkg/cli.doMain(0x7d4d7a0, 0xc0017a150a, 0x10, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:121 +0x11a
github.com/cockroachdb/cockroach/pkg/cli.Main()
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:59 +0x13a
main.main()
	/go/src/github.com/cockroachdb/cockroach/pkg/cmd/cockroach/main.go:26 +0x25
```

I would expect this not to crash, but either provide a human-suitable error message or do whatever the command is supposed to do.

**Comment by ajwerner** [comment]
Can you please provide the versions for the server and cli in use?

**Comment by smcvey** [comment]
21.1.8 both server and cli.

**Comment by cpnuj** [comment]
You can try to add flag `--replicated` to your command, like `cockroach debug range-data cockroach-data 37 --replicated`

It would show you the replicated data in range 37. I think there might be some bugs while reading field `raftpb.HardState.term` which belongs to the unreplicated range data

**Comment by smcvey** [comment]
Confirmed that `--replicated` does work.

**Comment by andreimatei** [comment]
Saw this happen again, I believe with a 21.1.10 binary. @erikgrinaker , I think I remember you fixing something in the dump command. Does this sound familiar?

**Comment by erikgrinaker** [comment]
Yeah, it was fixed in #63299, but doesn't look like it was backported to 21.1 because it pulled in a bunch of gogoproto upgrades, which is a bit risky.

**Comment by erikgrinaker** [comment]
Confirmed that @smcvey's repro above works with 21.2, so going to close this.
